pkg/alertdog/alertdog.go

```diff
@@ -0,0 +1,161 @@
+package alertdog
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/PagerDuty/go-pagerduty"
+	"github.com/prometheus/alertmanager/template"
+
+	"github.com/errm/alertdog/pkg/alertmanager"
+)
+
+type Alertmanager interface {
+	Alert(alertmanager.Alert) error
+	Resolve(alertmanager.Alert) error
+}
+
+type Pagerduty interface {
+	ManageEvent(event pagerduty.V2Event) (*pagerduty.V2EventResponse, error)
+}
+
+type Alertdog struct {
+	AlertmanagerEndpoints []string `yaml:"alertmanager_endpoints"`
+	Expected              []*Prometheus
+	CheckInterval         time.Duration
+	Expiry                time.Duration
+	PagerDutyKey          string
+	PagerDutyRunbookURL   string
+
+	mu           sync.Mutex
+	checkedIn    time.Time
+	alertmanager Alertmanager
+	pagerduty    Pagerduty
+}
+
+func (a *Alertdog) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	defaultInterval, _ := time.ParseDuration("2m")
+	a.CheckInterval = defaultInterval
+	defaultExpiry, _ := time.ParseDuration("5m")
+	a.Expiry = defaultExpiry
+	type plain Alertdog
+	return unmarshal((*plain)(a))
+}
+
+func (a *Alertdog) Setup() {
+	a.alertmanager = alertmanager.Alertmanager{Endpoints: a.AlertmanagerEndpoints, Expiry: a.CheckInterval * 2}
+	a.pagerduty = PagerdutyClient{}
+}
+func (a *Alertdog) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	var data template.Data
+	if err := json.NewDecoder(r.Body).Decode(&data); err != nil {
+		log.Printf("Webhook body invalid, skipping request: %s", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	for _, alert := range data.Alerts {
+		a.processWatchdog(alert)
+	}
+	w.WriteHeader(http.StatusOK)
+	return
+}
+
+func (a *Alertdog) processWatchdog(alert template.Alert) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
```

Comment:
❓ Is acquiring lock here not to conflict with [`CheckLoop()`](https://github.com/errm/alertdog/pull/1/files#diff-3ff4a69d33eee30968403b3f72c8b73c2ba195440f265dfa1c421a169fee4cedR87-R113) ? Which path do we specifically care about the lock contention? Can we narrow down the scope of mutex lock here (e.g. before/after triggering Alertmanager API within for loop)

Reply:
I specificly don't want a webhook request to be updating a promethue.CheckIn at the same time we are reading from it.  Or for multiple requests to be updating the same variable at once.

Thinking about it... it might be better to have a RW mutex on each prometheus object.

Reply:
I pushed a new commit... does this look a little better to you @kenju ?

Reply:
@errm Looks much better now. Now it is obvious where we want to acquire locks. Thanks.